### DISCUSSION TITLE
Metadata on output.file.hls shows only title.

### DISCUSSION THREAD

**Discussion by exlexv** [savonet/liquidsoap#3324] [User Support]
Hello! Using latest  2.2.0 version via opam. 

Everything works great except one thing. My radio setup outputs both to Icecast and HLS. So when I connect to Icecast source `example.com/stream` with my phone app (audio stream player) I can see both current 'artist' and 'title' on my screen. But when I connect to HLS source `example.com/hls/stream.m3u8` — I see only song 'title'. How can I fix this?

The strange thing is
I've tested: if I take the source and make the output only to 'Icecast' and then take the Icecast url as an input to output.file.hls — I get both 'artist' and 'title' again.

>url = "http://example:8000/stream"
rock = mksafe(input.http(url))
aac_lofi = %ffmpeg(format="mpegts",%audio( codec="aac",channels=2, ar=44100,b="128k"))
aac_midfi = %ffmpeg(format="mpegts",%audio(codec="aac",channels=2, ar=44100,b="192k")) 
streams_rock = [("rock_lofi",aac_lofi),("rock_midfi", aac_midfi)]
def segment_name(~position,~extname,stream_name) =
  timestamp = int_of_float(time())
  duration = 3
  "#{stream_name}_#{duration}_#{timestamp}_#{position}.#{extname}"
end
output.file.hls(playlist="stream.m3u8",
                segment_duration=3.0,
                segments=10,
                segments_overhead=10,
                segment_name=segment_name,
                persist_at="/var/www/html/hls/state.config",
                "/var/www/html/hls",
                streams_rock,
                rock)`

This way I get both 'artist' and 'title' from HLS output.

And this way I get only 'title' on HLS output but 'artist' and 'title' on Icecast:

> rock = mksafe(playlist(reload_mode="watch", "playlist/playlist.m3u"))
output.icecast(%mp3(bitrate=128), host = "example.com", port = 8000, password = "secret", mount = "stream", rock)
aac_lofi = %ffmpeg(format="mpegts",%audio( codec="aac",channels=2, ar=44100,b="128k"))
aac_midfi = %ffmpeg(format="mpegts",%audio(codec="aac",channels=2, ar=44100,b="192k")) 
streams_rock = [("rock_lofi",aac_lofi),("rock_midfi", aac_midfi)]
def segment_name(~position,~extname,stream_name) =
  timestamp = int_of_float(time())
  duration = 3
  "#{stream_name}_#{duration}_#{timestamp}_#{position}.#{extname}"
end
output.file.hls(playlist="stream.m3u8",
                segment_duration=3.0,
                segments=10,
                segments_overhead=10,
                segment_name=segment_name,
                persist_at="/var/www/html/hls/state/state.config",
                "/var/www/html/hls",
                streams_rock,
                rock)

**Comment by toots** [comment]
Hi @exlexv ! The reason for this is that, historically, icecast has a special `song` metadata that is filled with `#{title} - #{artist}`

If title is the metadata being displayed from the hls stream, you can emulate this behavior by using `metadata.map` as follows:

```liquidsoap
# Update title metadata when both artist and title are present.
def hls_meta(m) =
  title = m["title"]
  artist = m["artist"]
  if artist != "" and title != "" then
    [("title", "#{artist} - #{title}")]
  else
    []
  end
end

hls_source = metadata.map(source, hls_meta)
```

**Reply by exlexv** [reply]
Hi @toots ! Thanks for your reply. I did what you suggest before asking this question. It solves the HLS output meta for sure, but along the way it changes Icecast output data as well and on Icecast output I get: `Artist - Artist - Title`

So if I do:

> my_icecast_source = mksafe(playlist(reload_mode="watch", "playlist/playlist.m3u"))
output.icecast(%mp3(bitrate=128), host = "example.com", port = 8000, password = "secret", mount = "stream", my_icecast_source)
...
...
def hls_meta(m) =
  title = m["title"]
  artist = m["artist"]
  if artist != "" and title != "" then
    [("title", "#{artist} - #{title}")]
  else
    []
  end
end
hls_source = metadata.map(hls_meta, my_icecast_source)
...
output.file.hls(playlist="rock.m3u8",
                segment_duration=3.0,
                segments=10,
                segments_overhead=10,
                segment_name=segment_name,
                persist_at="/var/www/html/state.config",
                "/var/www/html/hls",
                streams_rock,
               hls_source)

I get on Icecast:  `Artist - Artist - Title`
HLS:  `Artist - Title`

Logically If create new source for HLS (`hls_source = metadata.map(hls_meta, my_icecast_source)`) it should apply only to `hls_source` but instead it changes `my_icecast_source` source as well. So what should be done in order not to change Icecast meta along the way?

**Reply by exlexv** [reply]
Hmm, this script started working fine when I excluded `clock.assign_new([my_icecast_source])` which i was having in it

**Reply by exlexv** [reply]
So, actually it is getting stranger and stranger)) I included `clock.assign_new([my_icecast_source])` again and script started working fine. Probably it has something to do with clocks. And sometimes first song starts playing with Icecast `Artist - Artist - Title` and then everything goes back to normal. But then again `Artist - Artist - Title` behavior appears.

I have 4 outputs right now: 2 Icecast and 2 HLS:
First_Radio is divided into 2 sources (Icecast (First_Radio_1) and HLS (First_Radio_2 - metadata change)), 

Second_Radio is divided into 2 sources (Icecast (Second_Radio_1) and HLS (Second_Radio_2 - metadata change)).

So now how u should assign clocks to different sources in a right way?

**Reply by toots** [reply]
You should apply that mapping only to the source that feeds into the HLS output!

**Comment by exlexv** [comment]
@toots 
Did as you have said and have been testing all this time. Took `my_icecast_source' which goes to output.icecast and made a new source for HLS:

`hls_source = metadata.map(hls_meta, my_icecast_source)`

This resulted in two outputs:

1. `output.icecast(..., my_icecast_source)`
2. `output.file.hls(...,hls_source)`

Result: 
metadata from `output.file.hls(...,hls_source)` : 'Artist' - 'Title'

metadata from `output.icecast(..., my_icecast_source)` :      30% / 70%      —     'Artist' - 'Title' / 'Artist' - 'Artist'  - 'Title'

So as you can tell mapping has been applied only to `hls_source` but nevertheless it does some tricks to `my_icecast_source` as well (30% / 70%). Sometimes it doesn't affect it, mostly it does.

**Reply by exlexv** [reply]
As I've been writing this - more time has passed. Result is the same on metadata from output.icecast:

30-40% is ok ( 'Artist' - 'Title')

60-70% are affected by mapping for another hls_source ('Artist' - 'Artist' - 'Title')

**Reply by toots** [reply]
Hmm yes we've had mutable metadata historically, which is annoying and fixed in `main`. Can you try this instead:

```liquidsoap
# Update title metadata when both artist and title are present.
def hls_meta(m) =
  title = m["title"]
  artist = m["artist"]
  if artist != "" and title != "" then
    [...list.assoc.remove("title", m), ("title", "#{artist} - #{title}")]
  else
    m
  end
end

hls_source = metadata.map(update=false, source, hls_meta)
```

**Reply by exlexv** [reply]
The result is the same. (I changed places of `source` and `hls_meta` - `hls_source = metadata.map(update=false, hls_meta, source)` otherwise it won't start). But the result stays the same.
hls_source - OK!
source - 30% / 70% success

I have described it more clearly here - https://github.com/savonet/liquidsoap/issues/3325 Already tried different options, but nothing comes to mind.

**Reply by toots** [reply]
Have you tried the updated function above? This is a different version than the one I suggested before.

**Reply by exlexv** [reply]
Yes, I'm running script with it right now. Result is the same.  

>icecast_rockradio = rockradio
output.icecast(%mp3(bitrate=128), ........, icecast_rockradio)
def hls_meta_rock(m) =
  title = m["title"]
  artist = m["artist"]
  if artist != "" and title != "" then
    [...list.assoc.remove("title", m), ("title", "#{artist} - #{title}")]
  else
    m
  end
end
HLS_rockradio = metadata.map(update=false, hls_meta_rock, rockradio)
output.file.hls(playlist="rock.m3u8",
                segment_duration=3.0,
                segments=10,
                segments_overhead=10,
                segment_name=segment_name,
                persist_at="/var/www/html/hls/state.config",
                "/var/www/html/omfm/hls",
                streams_rock,
                HLS_rockradio)

Here are screenshots:

1. From icecast:

 
<img width="500" alt="Снимок экрана 2023-08-16 в 16 14 38" src="https://github.com/savonet/liquidsoap/assets/108526222/87a9649e-8a13-43ea-b6ae-c20b57f8976e">


2. From app which receives HLS stream

<img width="300" alt="Снимок экрана 2023-08-16 в 16 14 38" src="https://github.com/savonet/liquidsoap/assets/108526222/e651109d-67fe-461a-9161-e0970020b54a">

Recompiled liquidsoap and made script crystal clear and result the same:

>source = mksafe(source.m3u)
output.icecast(%mp3(bitrate=128), ........, source)
...
def hls_meta(m) =
  title = m["title"]
  artist = m["artist"]
  if artist != "" and title != "" then
    [...list.assoc.remove("title", m), ("title", "#{artist} - #{title}")]
  else
    m
  end
end
HLS_source = metadata.map(update=false, hls_meta, source)
output.file.hls(playlist="stream.m3u8",
                segment_duration=3.0,
                segments=10,
                segments_overhead=10,
                segment_name=segment_name,
                persist_at="/var/www/html/hls/state.config",
                "/var/www/html/hls",
                streams_aac,
                HLS_source)

**Reply by exlexv** [reply]
So... what i did is:
1. Modified your last function — changed title to artist to remove artist from metadata:
>def hls_meta(m) =
  title = m["title"]
  artist = m["artist"]
  if artist != "" and title != "" then
    [...list.assoc.remove(`"artist"`, m), ("title", "#{artist} - #{title}")]
  else
    m
  end
end
hls_source = metadata.map(update=false, hls_meta, source)
output.file.hls(playlist="stream.m3u8",hls_source)
output.icecast(...,source)

Hope it makes sense. Now everything works.  But again, if you add something to `"#{artist} - #{title} - example.com"` to modify hls output it will affect output.icecast as well. So something needs to be done to be able to use different metadata for output.file.hls and output.icecast. With the above function I see no difference in metadata outputs for both icecast and HLS, but if I change `"#{artist} - #{title}"` to `"#{artist} - #{title} - example.com"` I would see this `example.com` on output.icecast in 70% cases.

So you either use the same meta on  output.file.hls and output.icecast or otherwise it won't work.



2. Another option - make the same metadata for both outputs
 >def `hls_and_icecast_meta(m)` =
  title = m["title"]
  artist = m["artist"]
  if artist != "" and title != "" then
    [...list.assoc.remove(`"artist"`, m), ("title", "#{artist} - #{title}")]
  else
    m
  end
end
source = metadata.map(update=false, hls_and_icecast_meta, source)
output.file.hls(playlist="stream.m3u8",source)
output.icecast(...,source)

This way works too cause metadata turns out to be identical. But if you use lastfm.submit you should take as a source unmodified meta like:
>`source.on_metadata(lastfm.submit(...))` 
source_final = metadata.map(update=false, hls_and_icecast_meta, source)
output.file.hls(playlist="stream.m3u8",source_final)
output.icecast(...,source_final)

Both above solutions are not "clean" solutions and something needs to be done to make it possible to use different metadata for different outputs because today I couldn't make it work as intended but got to the pointt when both metadatas are identical. Somebody might want to inject something into HLS stream for example and keep original metadata for Icecast output. But now it is unachievable and if one tries to modify output.file.hls metadata it would affect output.icecast metadata in 70% of cases (tracks).

Hope I'm making myself clear, sorry for my english.

**Comment by Misheck0** [comment]
i have issue with liquidsoap to get song title and artist . here is route Route::get('/playlist.m3u', function () {
    try {
        $songs = Songs::orderBy('created_at', 'desc')->get();

        $lines = $songs->map(function ($song) {
            $title = str_replace(['"', "'", "\n", "\r"], '', $song->title ?? 'Unknown Title');
            $artist = str_replace(['"', "'", "\n", "\r"], '', $song->artist ?? 'Unknown Artist');
            $url = url($song->path);

            $annotateLine = "annotate:title=\"{$title}\",artist=\"{$artist}\":{$url}";
            $extinfLine = "#EXTINF:-1,{$artist} - {$title}\n{$url}";

            return $annotateLine . "\n" . $extinfLine;
        });

        $content = "#EXTM3U\n" . $lines->implode("\n");

        Log::info("Generated playlist.m3u with annotate:\n" . $content);

        return response($content, 200)
            ->header('Content-Type', 'audio/x-mpegurl')
            ->header('Cache-Control', 'no-cache, must-revalidate');

    } catch (\Exception $e) {
        Log::error("Playlist generation failed: " . $e->getMessage());
        return response("Error generating playlist", 500);
    }
});
the song_title,artist,song_path are been stored in the table. on my radio.liq #!/usr/bin/liquidsoap
set("log.file.path", "/home/misheck/liquidsoap.log")

# Fetch playlist from Laravel route
playlist_url = "http://127.0.0.1:8001/playlist.m3u"
playlist_source = playlist(playlist_url, reload=60)

# Wrap in safe and add crossfade
playlist_radio = mksafe(crossfade(playlist_source))

# Live stream input
live = input.http("http://127.0.0.1:8001/liveSongs")

# Fallback between live and playlist
main = fallback(track_sensitive=false, [live, playlist_radio])



# Log current song and send to Laravel API
def log_and_notify(m) =
  

  title = m["title"]
  artist = m["artist"]
  log("🎵 Now playing: #{artist} - #{title}")

  curl_command = "curl -X POST http://127.0.0.1:8001/api/now-playing " ^
                 "-H 'Content-Type: application/json' " ^
                 "-d '{\"title\": \"" ^ title ^ "\", \"artist\": \"" ^ artist ^ "\"}'"
  ignore(get_process_lines(curl_command))

  [("title", title), ("artist", artist)]
end

# Attach metadata function to playlist stream only
main = map_metadata(log_and_notify, main)

# Icecast Output with metadata
output.icecast(
  %mp3,
  mount = "/liveSongs",
  host = "localhost",
  port = 8000,
  password = "",
  name = "My Laravel Radio",
  description = "Streaming from Laravel",
  url = "http://127.0.0.1:8001/",
  genre = "Various",
  icy_metadata="true",
  main
)
 log_and_notify(m)  function need to get the artist name and title then send a post request to the api but once the post request sent the logs are empty    Log::info('Now Playing Payload22:', $request->all());
                       Route::post('/now-playing', function (Request $request) {
     Log::info('Now Playing Payload22:', $request->all());
    // Sanitize input
    $title = $request->input('title');
    $artist = $request->input('artist');
    //log

    // Reset all songs
    Songs::where('now_playing', true)->update(['now_playing' => false]);

    // Try to find the current song
    $song = Songs::where('title', $title)
                ->when($artist, fn($query) => $query->where('artist', $artist))
                ->first();

    if ($song) {
        $song->now_playing = true;
        $song->save();
        return response()->json(['message' => 'Now playing updated']);
    }

    return response()->json(['message' => 'Song not found'], 404);
});
